### PR TITLE
LTD-4828-add-registration-django-forms-summary-page 

### PR DESCRIPTION
**Pull request by depsiatwal**
https://uktrade.atlassian.net/browse/LTD-4828

Adds a new confirmation screen the the Django Registration flow. 
Has the ability to go back wizard steps to edit fields.

### CODE REVIEW

File: exporter/core/organisation/views.py [hunk 1]
```diff
@@ -97,6 +102,33 @@ def get_registration_address_form(self, data):
...
+        template_name = super().get_template_names()
+        if self.steps.current == RegistrationSteps.REGISTRATION_CONFIRMATION:
+            template_name = [self.template_list[self.steps.current]]
+        return template_name
+
+    def get_context_data(self, form, **kwargs):
+        context = super().get_context_data(form, **kwargs)
+        if self.steps.current == RegistrationSteps.REGISTRATION_CONFIRMATION:
+            form_dict = OrderedDict()
+            for form_key in self.get_form_list():
```
Comment (depsiatwal) [inline exporter/core/organisation/views.py:115]: This was done this way as get_all_cleaned_data()  required unique form fields. As duplicates get overridden. Hence the Payloadbuilder is used to get our data.

File: unit_tests/exporter/core/organisation/test_organisation_views.py [hunk 2]
```diff
@@ -155,6 +156,25 @@ def test_registration_individual_end_to_end_uk_based(
...
+    assert list(response.context["registration_data"].keys()) == [
+        "type",
+        "location",
+        "name",
+        "eori_number",
+        "vat_number",
+        "site",
+        "phone_number",
+        "website",
+    ]
```
Comment (kevincarrogan) [inline unit_tests/exporter/core/organisation/test_organisation_views.py:172]: Why not check the actual values here?

File: unit_tests/exporter/core/organisation/test_organisation_views.py [hunk 3]
```diff
@@ -155,6 +156,25 @@ def test_registration_individual_end_to_end_uk_based(
...
+
+    assert response.context["is_individual"] == True
+    assert response.context["is_uk_based"] == True
```
Comment (kevincarrogan) [inline unit_tests/exporter/core/organisation/test_organisation_views.py:175]: `assert response.context["is_individual"]` is enough, you don't need to equate this to `True` as it's redundant.

File: unit_tests/exporter/core/organisation/test_organisation_views.py [hunk 4]
```diff
@@ -209,6 +229,22 @@ def test_registration_individual_end_to_end_non_uk_based(
...
+        "name",
+        "eori_number",
+        "vat_number",
+        "site",
+        "phone_number",
+        "website",
+    ]
+
+    assert response.context["is_individual"] == True
+    assert response.context["is_uk_based"] == False
```
Comment (kevincarrogan) [inline unit_tests/exporter/core/organisation/test_organisation_views.py:244]: `assert not response.context["is_uk_based"]` is enough here. You don't need to equate it to False.

File: exporter/templates/core/registration/confirmation-registration.html [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	<h1 class="govuk-heading-m">
+		{% if is_individual and is_uk_based %}
+			Where in the United Kingdom are you based?
+		{% elif is_individual %}
+			Where are you based?
+		{% else %}
+			{% if is_uk_based %}
+				What is your registered office address?
+			{% else %}
+				Where is your orgainsation based ?
```
Comment (hnryjmes) [inline exporter/templates/core/registration/confirmation-registration.html:35]: This probably should be `Where is your organisation based?`

Comment (depsiatwal) [reply exporter/templates/core/registration/confirmation-registration.html]: corrected

File: unit_tests/exporter/core/organisation/test_organisation_views.py (hunk 3 above)
Comment (depsiatwal) [reply unit_tests/exporter/core/organisation/test_organisation_views.py]: done

File: unit_tests/exporter/core/organisation/test_organisation_views.py (hunk 2 above)
Comment (depsiatwal) [reply unit_tests/exporter/core/organisation/test_organisation_views.py]: done

File: exporter/core/organisation/views.py [hunk 6]
```diff
@@ -97,6 +102,33 @@ def get_registration_address_form(self, data):
...
+    def get_template_names(self):
+        template_name = super().get_template_names()
+        if self.steps.current == RegistrationSteps.REGISTRATION_CONFIRMATION:
+            template_name = [self.template_list[self.steps.current]]
+        return template_name
+
+    def get_context_data(self, form, **kwargs):
+        context = super().get_context_data(form, **kwargs)
+        if self.steps.current == RegistrationSteps.REGISTRATION_CONFIRMATION:
+            form_dict = OrderedDict()
```
Comment (saruniitr) [inline exporter/core/organisation/views.py:114]: I am not able to follow why special handling is required for this confirmation screen.
When adding a product also we display summary screen with change links,
eg https://github.com/uktrade/lite-frontend/blob/dev/exporter/applications/views/goods/firearm/views/summary.py
and my understanding is it is a similar pattern here.

File: exporter/templates/core/registration/confirmation-registration.html [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+</div>
+<form action="" method="post"{% if form.helper.attrs.enctype %} enctype="{{ form.helper.attrs.enctype }}"{% endif %}>
+	{% csrf_token %}
+	{{ wizard.management_form }}
+
+	{% if is_individual %}
+		{% include "core/registration/includes/registration-details-individual.html" %}
+	{% else %}
+		{% include "core/registration/includes/registration-details-corporation.html" %}
+	{% endif %}
```
Comment (saruniitr) [inline exporter/templates/core/registration/confirmation-registration.html:24]: Multiple if conditions are making it hard to follow, as we are already including additional templates based on if conditions, could it be refactored to include all additional templates in just one template so this one will just have one `if-else` branch rather than for each question?

or even better if we conditionally update the template in the view itself. We have separate templates for individual/commercial, uk based/outside uk and the template name is dynamically specified in the view.

File: exporter/core/organisation/views.py (hunk 6 above)
Comment (depsiatwal) [reply exporter/core/organisation/views.py:114]: nah this is a different pattern. The product has change links and the product gets saved to the DB and we present the ability to make changes after. 

This confirmation screen presents all the data to the user before we submit to DB

File: exporter/templates/core/registration/confirmation-registration.html (hunk 7 above)
Comment (depsiatwal) [reply exporter/templates/core/registration/confirmation-registration.html:24]: Yeah I did think about this for a while. 
The first set is used just to change the title of the form and the 2nd changes what's displayed. 

since very little changes between UK/non-UK based I was very reluctant to specify different templates since we only change the title.

Comment (depsiatwal) [reply exporter/templates/core/registration/confirmation-registration.html]: Done move to individual templates

File: exporter/templates/core/registration/includes/address-details-abroad.html [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        <dd class="govuk-summary-list__value">
+            {{ registration_data.website}}
+        </dd>
+        <dd class="govuk-summary-list__actions">
+            <button name="wizard_goto_step" type="submit" value="ADDRESS_DETAILS" class="lite-button--link">Change</button>
+        </dd>
+    </div>
+    <div class="govuk-summary-list__row" id="org-country">
+        <dt class="govuk-summary-list__key">
+            County
```
Comment (saruniitr) [inline exporter/templates/core/registration/includes/address-details-abroad.html:49]: typo in country

**Review by saruniitr — APPROVED** [review]
with the typos fixed, it looks good to me

File: exporter/templates/core/registration/includes/address-details-uk.html [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        <dd class="govuk-summary-list__value">
+            {{ registration_data.site.address.city}}
+        </dd>
+        <dd class="govuk-summary-list__actions">
+            <button name="wizard_goto_step" type="submit" value="ADDRESS_DETAILS" class="lite-button--link"> Change</button>
+        </dd>
+    </div>
+    <div class="govuk-summary-list__row" id="org-country">
+        <dt class="govuk-summary-list__key">
+            County or state
```
[the remark this answers is not in the document]
Comment (depsiatwal) [reply exporter/templates/core/registration/includes/address-details-uk.html:39]: it's weird it's kind of what they summaries on the current application. 
they may change this during product review.

File: exporter/templates/core/registration/includes/address-details-abroad.html (hunk 8 above)
Comment (depsiatwal) [reply exporter/templates/core/registration/includes/address-details-abroad.html:49]: fixed

**Review by hnryjmes — APPROVED** [review]
